fix: issues #1206 and #1208 - unique IDs and loading state

Jhonattan2121 opened the pull request:
# Changes Implemented - Issues #1206 and #1208

## Issue #1206: Generate unique fidget IDs

**Problem:** Human-readable IDs caused collisions when multiple fidgets were created.

**Solution:** Implemented automatic UUID generation for each fidget when space config is received from vibe customizer.

**What was done:**
- Each fidget receives a unique UUID (`uuidv4()`)
- Same ID is applied to both `fidgetInstanceDatums` and `layout`
- Implemented in two entry points: `AgentChat.tsx` and `ThemeSettingsEditor.tsx`

**Files modified:**
- `src/common/components/organisms/AgentChat.tsx`
- `src/common/lib/theme/ThemeSettingsEditor.tsx`

---

## Issue #1208: Loading state during customization

**Problem:** Users could modify the space while agents were customizing, causing conflicts.

**Solution:** Implemented loading overlay that blocks interactions during the building process.

**What was done:**
- Loading state added to `chatStore` (isLoading, loadingType)
- Visual overlay with spinner and "Building your space..." message
- Complete interaction blocking during building
- Loading remains active until configuration is fully applied

**Files modified:**
- `src/common/data/stores/app/chat/chatStore.ts`
- `src/common/components/organisms/AgentChat.tsx`
- `src/app/(spaces)/Space.tsx`

---

## Result

✅ **Issue #1206**: Unique IDs automatically generated  
✅ **Issue #1208**: Loading state prevents modifications during building


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

## Release Notes

* **New Features**
  * Added visual loading indicator with spinner that displays during space building operations, providing clear feedback on processing status

* **Improvements**
  * Enhanced loading state management across chat and space operations for better consistency
  * Improved error handling with clearer messaging when configuration resources are unavailable

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Jhonattan2121 committed:
fix: generate unique fidget IDs and add loading state for vibe customization

- Issue #1206: Generate unique UUIDs for each fidget when space config is received from vibe customizer
- Issue #1208: Display loading overlay on space while agents are building
- Add loading state to chat store (isLoading, loadingType)
- Keep loading active until autoApplyConfig completes
- Apply unique ID generation in both AgentChat and ThemeSettingsEditor
- Fix TypeScript errors in runtimeLoader and ThemeSettingsEditor

willyogo commented:
@Jhonattan2121 i'm not seeing the loading state while vibe customization is in progress

<img width="1505" height="856" alt="image" src="https://github.com/user-attachments/assets/5e661c02-7338-47a6-9f61-2b0d02ced5d8" />

Jhonattan2121 committed:
fix: ensure fidgets stay within grid boundaries and fix loading state

- Fix fidgets being generated outside grid boundaries by AI
  - Add comprehensive boundary validation in gridCleanup.ts
  - Ensure item dimensions don't exceed grid size (cols/maxRows)
  - Reposition out-of-bounds items automatically within grid
  - Add render-time validation in Grid.tsx to prevent rendering out-of-bounds items

- Fix incorrect hasProfile/hasFeed detection
  - Use new layout data instead of old context when applying AI config
  - Ensures correct grid dimensions (6/12 cols, 8/10 rows) are used

- Fix loading state getting stuck on 'Thinking...'
  - Add 60-second timeout to automatically clear stuck loading state
  - Clear loading state on WebSocket disconnect/error
  - Clear loading state on component unmount
  - Clear any stuck loading state on component mount
  - Properly clean up timeouts when responses are received

- Apply cleanup before layout is applied
  - Run comprehensiveCleanup in autoApplyConfig before applying layout
  - Ensures all fidgets are validated and repositioned before rendering

Files modified:
- src/common/components/organisms/AgentChat.tsx
- src/common/lib/theme/ThemeSettingsEditor.tsx
- src/common/lib/utils/gridCleanup.ts
- src/fidgets/layout/Grid.tsx

Jhonattan2121 committed:
fix: improve loading state during vibe customization

- Add loadingMessage field to chat store to track current activity
- Display dynamic loading messages in Space overlay and chat sidebar
- Implement gap timeout mechanism to show 'Processing...' after 3s of inactivity
- Add interval to rotate loading messages during long gaps (every 5s)
- Increase timeout from 60s to 10 minutes for building operations
- Protect building mode from premature timeout clearing
- Keep loading active until config is fully applied
- Remove console.logs as requested
- Update loading message immediately when config arrives

Fixes issue where loading would stop during gaps between builder logs
and final config delivery, providing continuous feedback to users.

Jhonattan2121 committed:
feat: auto-fill grid with fidgets and enforce strict boundary validation

- Add fillEmptySpaces function to automatically generate fidgets in empty grid spaces
- Implement multi-layer strict validation to ensure fidgets never exceed grid boundaries
- Add final validation in comprehensiveCleanup to remove invalid fidgets
- Add final validation in fillEmptySpaces before returning results
- Add extra validation in AgentChat and ThemeSettingsEditor before filling spaces
- Add final validation before saving configuration to ensure no invalid fidgets are persisted
- Fix issue where fidgets were rendered outside grid boundaries causing console warnings
- Limit additional fidget generation (4-6 depending on grid size) to prevent overcrowding
- Automatically stop when grid is completely occupied

Fixes: fidgets being generated outside grid boundaries
Fixes: grid not being automatically filled when there are empty spaces

Jhonattan2121 commented:
<img width="1653" height="1102" alt="Screenshot From 2025-12-17 20-11-22" src="https://github.com/user-attachments/assets/013f3111-387a-4600-91c3-db3c79624460" />
@willyogo  @j-paterson 

The fixes have been implemented 
 now the AI generates widgets within the grid without spilling outside of it.

willyogo commented:
@Jhonattan2121 it's not just that #1620 combines two issues, but three. 
- fidget uuids
- loading state (PR description says this is part of it, but it's not working)
- and now grid stuff (initial issue was just to pass the grid size to the vibe coding agent; i'm not sure this is the approach we want to take, and considering the current quality of the vibe customization feature, am now thinking we may want to hide it for now until we're able to prioritize it and make it work better (plus give time for models to improve))

for now, please pause all work on vibe customization stuff and take a look at this bug: https://github.com/blankdotspace/nounspace.ts/issues/1623
